sycl/include/sycl/ext/intel/experimental/kernel_execution_properties.hpp

```diff
@@ -26,12 +26,19 @@ inline constexpr cache_config_enum large_slm =
 inline constexpr cache_config_enum large_data =
     cache_config_enum::large_data;
 
-struct cache_config {
-  cache_config(cache_config_enum v) : value(v) {}
-  cache_config_enum value;
+struct cache_config_key {
+  using value_t = oneapi::experimental::property_value<cache_config_key>;
 };
-
-using cache_config_key = cache_config;
+using cache_config = cache_config_key::value_t;
+} // namespace ext::intel::experimental
+namespace ext::oneapi::experimental {
+template <> struct property_value<intel::experimental::cache_config_key> {
+  using key_t = intel::experimental::cache_config_key;
+  property_value(intel::experimental::cache_config_enum v) : value(v) {}
+  intel::experimental::cache_config_enum value;
+};
+} // namespace ext::oneapi::experimental
+namespace ext::intel::experimental {
```

Comment:
You can see here that the definition of a typical runtime property expanded from 5 lines to about 12.  However, the description of the PR says that a goal is to simplify the implementation.  What is getting simplified?  It seems like the definition of a runtime property is more complex, not less.

Reply:
> What is getting simplified?

The main motivation is to simplify has_property and get_property. I'm worried that if/when properties are getting used a lot, that the current implementation will be bad for compilation speed. Currently we store the properties as a sorted list of values in the template argument of `properties`. And because not all have the same form (compile-time are of the form `property_value<key, ...>` and runtime are their own type) we have to iterate over the whole list every time.
With this change all property have the same form of `property_value<key, ...>`. This turns the sorted list into a sorted map. And with that efficient map lookup with mp_map_find ([doc](https://www.boost.org/doc/libs/1_83_0/libs/mp11/doc/html/mp11.html#mp_map_findm_k), [impl](https://github.com/boostorg/mp11/blob/develop/include/boost/mp11/detail/mp_map_find.hpp)) is possible.

If we don't like this change we could instead change the storage inside the template argument to a sorted set of runtime properties and a sorted map of compile-time properties (doesn't require any spec changes). Both would have fast lookup. We would just need to split the properties in the two group prior to sorting whenever we sort.

A 2nd motivation was to reduce where we need to distinguish between the two groups of properties in the implementation. With the alternative implementation this would become more not less.

For the spec we could allow both implementation strategies (runtime property values are of the form `property_value<key, ...>` or runtime property values/keys the same type). Then this becomes purely a quality of implementation question.

The `property_value<key, ...>` form allows a runtime property value to store some of its values as compile-time values. And those values can be queried at compile time. I'm not aware of use-cases for properties which have both runtime values (members) and also compile-time values (template args). So I don't know whether this is an advantage.

Reply:
I created a prototype of what I think a better implementation should roughly look like: https://godbolt.org/z/eM4cxK1bP. For the PRs I was trying to break the changes into incremental steps to help with review. But hopefully this shows what is possible after all changes. In my view it makes the implementation of the properties class much shorter and also easier to define both runtime and compile-time properties. This implementation currently isn't conformant, because the spec contains several implementation details which aren't valid for this strategy. To make it conformant we would need to remove (or change):
- Runtime properties key/value are the same type (this change)
- Remove key_t (https://github.com/intel/llvm/pull/12246, not necessary but makes defining runtime properties simpler )
- Remove static from `get_property` (not necessary but simplifies implementation and I don't think we have any use case which actual needs it to be static)

I would be interested in any feedback regarding:
- Do we want the spec to mandate one implementation strategy or not?
- Which implementation strategy do we want to use?
- Do you prefer for review a set of incremental changes or one large PR?

Reply:
I prefer multiple incremental PRs, personally, and I don't think we should specify anything about implementation that we don't have to.